### PR TITLE
Add ability to skip phases

### LINKED ISSUE #84: support triggering a later phase only and not require earlier taskgroups/phases
I believe the snowman/action-task logic supports this already but shipit itself may need to be told that this is something valid.

### PR DESCRIPTION
**Pull request by rail**
Fixes #84

### CODE REVIEW

File: api/src/shipit_api/api.py [hunk 1]
```diff
@@ -169,6 +169,12 @@ def schedule_phase(name, phase):
     phase.task_id = result["status"]["taskId"]
 
     phase.submitted = True
+    # Make sure that if one of the previous phases are either submitted or
+    # skipped
```
Comment (mozbhearsum) [inline api/src/shipit_api/api.py]: I'm having trouble understanding this comment. "if one of" makes me think there should be some additional clause at the end. Can you rephrase this?

File: api/src/shipit_api/api.py [hunk 2]
```diff
@@ -169,6 +169,12 @@ def schedule_phase(name, phase):
...
+    for ph in phase.release.phases:
+        if ph.name == phase.name:
+            break
+        ph.submitted = True
```
Comment (mozbhearsum) [inline api/src/shipit_api/api.py]: Just to make sure I understand this block correctly: It is marking all earlier phases as submitted -- right? It's going through the phases (which are ordered by id, if I understand the model correctly), and marking everything prior to the phase being submitted as "submitted".

File: api/src/shipit_api/models.py [hunk 3]
```diff
@@ -90,11 +90,18 @@ def rendered_hook_payload(self, extra_context={}):
...
+        # in case we skip a phase, the task ID is not defined, we want to
+        # filter it out
+        resolved_previous_graph_ids = filter(None, resolved_previous_graph_ids)
+        context["input"]["previous_graph_ids"] = list(resolved_previous_graph_ids)
         if extra_context:
             context.update(extra_context)
         return render_action_hook(self.task_json["hook_payload"], context)
 
+    @property
+    def skipped(self):
```
Comment (mozbhearsum) [inline api/src/shipit_api/models.py]: :+1: to having this be a calculated property, rather than a new column.

File: api/src/shipit_api/api.py (hunk 2 above)
Comment (mozbhearsum) [inline api/src/shipit_api/api.py]: Also, do `completed_by` or `completed` need to get set for the skipped phases, too?

File: api/src/shipit_api/api.py [hunk 4]
```diff
@@ -169,6 +169,12 @@ def schedule_phase(name, phase):
...
     phase.completed_by = current_user.get_id()
     completed = datetime.datetime.utcnow()
     phase.completed = completed
```
Comment (mozbhearsum) [inline api/src/shipit_api/api.py]: I noticed that `abandon_release` has some logic that loops over phases and tries to fetch artifacts for them. Is that going to work for skipped phases? I see that it catches `ArtifactNotFound`, but I'm not sure if that's what will be thrown for taskIds that don't exist.

File: frontend/src/views/ListReleases/index.js [hunk 5]
```diff
@@ -522,15 +534,17 @@ class TaskLabel extends React.PureComponent {
   };
 
   render() {
-    const { status, name, taskGroupUrl } = this.props;
-    if (status === 'blocked') {
+    const {
+      status, name, taskGroupUrl, tasksInProgress, allowPhaseSkipping,
+    } = this.props;
+    if ((status === 'blocked' && !allowPhaseSkipping) || status === 'skipped' || tasksInProgress) {
```
Comment (mozbhearsum) [inline frontend/src/views/ListReleases/index.js]: I'm having trouble evaluating this because I don't understand what `blocked` means as a phase status. Can you explain?

**Comment by rail** [comment]
> While looking to see if other code needs to be updated I noticed that `shipit_api.config.SIGNOFFS` is empty, and doesn't appear to be updated by anything. Is that intentional?

Yeah. The feature is implemented, but we don't use it for anything. So we left it like that to have an example of a signoff config.

File: api/src/shipit_api/api.py (hunk 1 above)
Comment (rail) [reply api/src/shipit_api/api.py]: How about "If the previous phases are not submitted, mark them as submitted and they will be calculated as skipped because they don't have taskId associated"

File: api/src/shipit_api/api.py (hunk 2 above)
Comment (rail) [reply api/src/shipit_api/api.py]: Correct. Probably I should add another `if` so I don't touch actually submitted phases and for the rest add a timestamp and the submitter as you suggested.

File: api/src/shipit_api/models.py (hunk 3 above)
Comment (rail) [reply api/src/shipit_api/models.py]: Heh, I actually took another take on this last week to see how it looks with a column holding this value. The DB migration part was a bit painful tbh. I'll use the current approach.

File: api/src/shipit_api/api.py (hunk 2 above)
Comment (rail) [reply api/src/shipit_api/api.py]: :+1:

File: api/src/shipit_api/api.py (hunk 4 above)
Comment (rail) [reply api/src/shipit_api/api.py]: Good point. It will work, but better to add the skipped ones to the filter, so we don't even try to cancel them.

File: frontend/src/views/ListReleases/index.js (hunk 5 above)
Comment (rail) [reply frontend/src/views/ListReleases/index.js]: In this context "blocked" means that you cannot submit it yet. The previous phase has to be submitted and the corresponding action task has to be done. If you submit too early, the action task may not find all the needed tasks from the previous graph. I rearranged some stuff here, hope it'll be a bit clearer.

File: api/src/shipit_api/models.py (hunk 3 above)
Comment (mozbhearsum) [reply api/src/shipit_api/models.py]: Regardless I think a calculated property makes sense. Otherwise we could end up in a state where taskId is not null, and skipped is true. Best to make it impossible to get into unknown states :)
